### ISSUE TITLE
Performance improvements to CPU profile code

### ISSUE THREAD

**Issue by mkustermann** [flutter/devtools#7917]
After having taken a look at the cpu profiling code in DevTools, we found:

* It unnecessarily encodes json maps and decodes them again: [devtools/pull/7916](https://github.com/flutter/devtools/pull/7916)
* It gets `Stream</* String | List<int> */>` from the websocket and then json decodes strings
  => It would be more efficient to avoid going from utf-8 to string and then to json and instead directly from utf8->json
  => This would probably require some refactoring in various packages, so maybe not easily doable?
* Way too many conversions
  * Bytes->String in websocket layer
  * String->JSON in `package:vm_service layer`
  * JSON -> service objects in `package:vm_service` layer (e.g. `vm_service.CpuSamples`, `vm_service.CpuSample`)
  * service objects -> `_CpuProfileTimelineTree` tree structure
     => Uses expandos (!!) to map service objects back to tree nodes
  * `_CpuProfileTimelineTree`  -> JSON "traceObject" (very expensive operation!)
  * Going from JSON "traceObject" back to normal objects in `CpuProfileData.fromJson` which builds json maps with strings and `CpuStackFrame` as values

It seems there's too many conversions from one data structure to another and the representations used for encoding this profiling information doesn't seem very efficient either.

**Comment by kenzieschmoll** [comment]
CC @bkonyi

**Comment by mkustermann** [comment]
Another thing is that I believe loading CPU profiles can freeze the UI for a while - most likely due to all this work happening synchronously without any yields to the event loop. I have observed 80 MB of json to be loaded for cpu profile, which may result in quite a bit more of in-memory json (consumed by maps, lists, etc - which aren't as compact as in the encoded form)  - possibly 100s of MBs it's churning through synchronously several times.

**Comment by kevmoo** [comment]
Could we make this function async? Then we can batch the processing and avoid janking frames...

**Comment by mkustermann** [comment]
Friendly ping. @kenzieschmoll @bkonyi Are there plans on improving this in the short/mid term?

**Comment by kenzieschmoll** [comment]
@mkustermann thanks for the performance analysis. It sounds like all of your recommendations would be great improvements. In practice, this isn't something we have the cycles to work on in the short term, but are more than happy to accept & review contributions if you'd like to take a stab at any of these.

**Comment by jakemac53** [comment]
Now that https://github.com/flutter/devtools/pull/8892 is merged, we are both much more efficient and also much more asynchronous.

There is still potentially some work that could be done, for instance the sorting still causes jank.

**Comment by kenzieschmoll** [comment]
@jakemac53 after the performance work you did on the CPU profiler back in February, which of the items in the issue description are resolved and which still remain?

**Comment by jakemac53** [comment]
I don't believe I directly addressed these suggestions, in the profiling I found other things to be the root cause. Very likely some of these would yield additional improvements but might not be worth doing as large CPU profiles can load pretty quickly now, although not instantaneously.

> * It unnecessarily encodes json maps and decodes them again: [devtools/pull/7916](https://github.com/flutter/devtools/pull/7916)

Already resolved it looks like

> * It gets `Stream</* String | List<int> */>` from the websocket and then json decodes strings
  => It would be more efficient to avoid going from utf-8 to string and then to json and instead directly from utf8->json
  => This would probably require some refactoring in various packages, so maybe not easily doable?

I looked into going directly from bytes to JSON (by fusing the utf8 and json encoders), but it was significantly slower in wasm mode.

> * Way too many conversions
>  * Bytes->String in websocket layer
>  * String->JSON in `package:vm_service layer`
>  * JSON -> service objects in `package:vm_service` layer (e.g. `vm_service.CpuSamples`, `vm_service.CpuSample`)
>  * service objects -> `_CpuProfileTimelineTree` tree structure
>     => Uses expandos (!!) to map service objects back to tree nodes
>  * `_CpuProfileTimelineTree`  -> JSON "traceObject" (very expensive operation!)
>  * Going from JSON "traceObject" back to normal objects in `CpuProfileData.fromJson` which builds json maps with strings and `CpuStackFrame` as values

I don't believe I did any of these.

**Comment by kenzieschmoll** [comment]
Adding additional information from @mkustermann:

The data structures used for this CPU profile collection and the processing code continues to be very inefficient:
=> It represents IDs as long strings (such as "isolates/4095163606998683-9440").
=> Many other strings (like uris, resolved uris) are encoded many times instead of having a canonicalized table and refer to them via integer ids
=> See [packages/devtools_test/lib/src/test_data/_cpu_profiler_data.dart](https://github.com/flutter/devtools/blob/master/packages/devtools_test/lib/src/test_data/_cpu_profiler_data.dart) (file content [here](https://raw.githubusercontent.com/flutter/devtools/refs/heads/master/packages/devtools_test/lib/src/test_data/_cpu_profiler_data.dart)) - the format is so redundant that gzip'ing it reduces size by -96.5%!
=> The verbose format means the json decoding is slow and the result will consume more memory.

=> When it builds up the cpu profiling tree it creates the same string id's over and over again (see [here](https://github.com/flutter/devtools/blob/e0925a8baf6394100dc8eb26f80a36c9b881b31e/packages/devtools_app/lib/src/screens/profiler/panes/method_table/method_table_model.dart#L153) and [here](https://github.com/flutter/devtools/blob/e0925a8baf6394100dc8eb26f80a36c9b881b31e/packages/devtools_app/lib/src/screens/profiler/panes/method_table/method_table_model.dart#L53)) - it could do that once per object.
=> This part is probably less efficient, because we have to re-create those strings, re-hash them, ... which is a little slower in dart2wasm than dart2js
